server/controllers/donations/addDonation.ts

```diff
@@ -0,0 +1,64 @@
+import { NextFunction, Response } from 'express';
+import { Campaign, Donation } from '../../database/models';
+import { CustomedError } from '../../utils';
+import { paramsSchema, donationSchema } from '../../utils/validation';
+
+const addDonation = async (req: any, res: Response, next: NextFunction) => {
+  try {
+    const {
+      user: { id: donorId },
+    } = req;
+    const {
+      params: { campaignId },
+    } = req;
+    const {
+      body: {
+        food,
+        clothes,
+        money,
+        description,
+        location,
+        deliver_time: deliverTime,
+      },
+    } = req;
+
+    await paramsSchema.validateAsync({ id: campaignId });
+
+    const campaign = await Campaign.findByPk(campaignId, {
+      raw: true,
+    });
+    if (!campaign) {
+      throw new CustomedError('Cannot add donation, campaign not exists', 400);
+    }
+
+    await donationSchema.validateAsync(req.body);
+
+    if (!food && !clothes && !money) {
+      next(
+        new CustomedError(
+          'You should enter money, piece of clothes, number of meals',
```

Comment:
i think this is better for a message what do you think ? "You should donate either money, clothes or meals"